core/arxiv/submission/services/filesystem/store.py

```diff
@@ -0,0 +1,112 @@
+"""
+Functions for storing and getting (not!) source packages from the filesystem.
+
+In the classic system we use a giant shared volume, which we'll call
+``LEGACY_FILESYSTEM_ROOT`` in this package. Inside of that, what we'll call a
+"shard id" (the first four digits of the submission ID) is used to create a
+directory that in turn holds a directory for each submission.
+
+For example, submission ``65393829`` would have a directory at
+``{LEGACY_FILESYSTEM_ROOT}/6539/65393829``.
+
+The submission directory contains a PDF that was compiled during the preview
+step of submission, a ``source.log`` file from the file management service
+(for the admins to look at), and a ``src`` directory that contains the actual
+submission content.
+
+We also require the ability to set permissions on files and directories, and
+set the owner user and group.
+
+To use this in a Flask application, the following config parameters must be
+set:
+
+- ``LEGACY_FILESYSTEM_ROOT``: (see above)
+- ``LEGACY_FILESYSTEM_SOURCE_DIR_MODE``: permissions for directories; see
+  :ref:`python:os.chmod`
+- ``LEGACY_FILESYSTEM_SOURCE_MODE``: permissions for files; see
+  :ref:`python:os.chmod`
+- ``LEGACY_FILESYSTEM_SOURCE_UID``: uid for owner user (must exist)
+- ``LEGACY_FILESYSTEM_SOURCE_GID``: gid for owner group (must exist)
+
+"""
+import os
+import tarfile
+import shutil
+
+from flask import current_app
+
+from .util import chmod_recurse, copy_with_mode
+
+
+class ConfigurationError(RuntimeError):
+    """A required parameter is invalid/missing from the application config."""
+
+
+class SecurityError(RuntimeError):
+    """Something suspicious happened."""
+
+
+def store_source(submission_id: int, source_path: str, pdf_path: str,
+                 log_path: str) -> None:
+    """
+    Deposit a source package into the legacy filesystem.
+
+    Parameters
+    ----------
+    submission_id : int
+        Numeric submission ID. This is **not** the upload ID.
+    source_path : str
+        Path to a gzipped tarball containing the source package to deposit.
```

Comment:
This implies that only tarballed source packages are supported. How about pdf-only or html submissions?